README.md
# NAME

GitLab::API::v4 - A complete GitLab API v4 client.

# SYNOPSIS

```perl
use GitLab::API::v4;

my $api = GitLab::API::v4->new(
    url           => $v4_api_url,
    private_token => $token,
);

my $branches = $api->branches( $project_id );
```

# DESCRIPTION

This module provides a one-to-one interface with the GitLab
API v4.  Much is not documented here as it would just be duplicating
GitLab's own [API Documentation](http://doc.gitlab.com/ce/api/README.html).

Note that this distribution also includes the [gitlab-api-v4](https://metacpan.org/pod/gitlab-api-v4) command-line
interface (CLI).

## Upgrading

If you are upgrading from [GitLab::API::v3](https://metacpan.org/pod/GitLab%3A%3AAPI%3A%3Av3) make sure you read:

[https://docs.gitlab.com/ce/api/v3\_to\_v4.html](https://docs.gitlab.com/ce/api/v3_to_v4.html)

Also, review the `Changes` file included in the distribution as it outlines
the changes made to convert the v3 module to v4:

[https://github.com/bluefeet/GitLab-API-v4/blob/master/Changes](https://github.com/bluefeet/GitLab-API-v4/blob/master/Changes)

Finally, be aware that many methods were added, removed, renamed, and/or altered.
If you want to review exactly what was changed you can use GitHub's compare tool:

[https://github.com/bluefeet/GitLab-API-v4/compare/72e384775c9570f60f8ef68dee3a1eecd347fb69...master](https://github.com/bluefeet/GitLab-API-v4/compare/72e384775c9570f60f8ef68dee3a1eecd347fb69...master)

Or clone the repo and run this command:

`git diff 72e384775c9570f60f8ef68dee3a1eecd347fb69..HEAD -- author/sections/`

## Credentials

Authentication credentials may be defined by setting either the ["access\_token"](#access_token)
or ["private\_token"](#private_token) arguments.

If no credentials are supplied then the client will be anonymous and greatly
limited in what it can do with the API.

Extra care has been taken to hide the token arguments behind closures.  This way,
if you dump your api object, your tokens won't accidentally leak into places you
don't want them to.

## Constants

The GitLab API, in rare cases, uses a hard-coded value to represent a state.
To make life easier the [GitLab::API::v4::Constants](https://metacpan.org/pod/GitLab%3A%3AAPI%3A%3Av4%3A%3AConstants) module exposes
these states as named variables.

## Exceptions

The API methods will all throw a useful exception if
an unsuccessful response is received from the API.  That is except for
`GET` requests that return a `404` response - these will return `undef`
for methods that return a value.

If you'd like to catch and handle these exceptions consider using
[Try::Tiny](https://metacpan.org/pod/Try%3A%3ATiny).

## Logging

This module uses [Log::Any](https://metacpan.org/pod/Log%3A%3AAny) and produces some debug messages here
and there, but the most useful bits are the info messages produced
just before each API call.

## Project ID

Note that many API calls require a `$project_id`.  This can be
specified as a numeric project `ID` or, in many cases, maybe all cases,
as a `NAMESPACE_PATH/PROJECT_PATH` string.  The GitLab documentation on
this point is vague.

# REQUIRED ARGUMENTS

## url

The URL to your v4 API endpoint.  Typically this will be something
like `https://git.example.com/api/v4`.

# OPTIONAL ARGUMENTS

## access\_token

A GitLab API OAuth2 token.  If set then ["private\_token"](#private_token) may not be set.

See [https://docs.gitlab.com/ce/api/#oauth2-tokens](https://docs.gitlab.com/ce/api/#oauth2-tokens).

## private\_token

A GitLab API personal token.  If set then ["access\_token"](#access_token) may not be set.

See [https://docs.gitlab.com/ce/api/#personal-access-tokens](https://docs.gitlab.com/ce/api/#personal-access-tokens).

## retries

The number of times the request should be retried in case it fails (5XX HTTP
response code).  Defaults to `0` (false), meaning that a failed request will
not be retried.

## sudo\_user

The user to execute API calls as.  You may find it more useful to use the
["sudo"](#sudo) method instead.

See [https://docs.gitlab.com/ce/api/#sudo](https://docs.gitlab.com/ce/api/#sudo).

## rest\_client

An instance of [GitLab::API::v4::RESTClient](https://metacpan.org/pod/GitLab%3A%3AAPI%3A%3Av4%3A%3ARESTClient) (or whatever ["rest\_client\_class"](#rest_client_class)
is set to).  Typically you will not be setting this as it defaults to a new
instance and customization should not be necessary.

## rest\_client\_class

The class to use when constructing the ["rest\_client"](#rest_client).
Defaults to [GitLab::API::v4::RESTClient](https://metacpan.org/pod/GitLab%3A%3AAPI%3A%3Av4%3A%3ARESTClient).

# UTILITY METHODS

## paginator

```perl
my $paginator = $api->paginator( $method, @method_args );

my $members = $api->paginator('group_members', $group_id);
while (my $member = $members->next()) {
    ...
}

my $users_pager = $api->paginator('users');
while (my $users = $users_pager->next_page()) {
    ...
}

my $all_open_issues = $api->paginator(
    'issues',
    $project_id,
    { state=>'opened' },
)->all();
```

Given a method who supports the `page` and `per_page` parameters,
and returns an array ref, this will return a [GitLab::API::v4::Paginator](https://metacpan.org/pod/GitLab%3A%3AAPI%3A%3Av4%3A%3APaginator)
object that will allow you to walk the records one page or one record
at a time.

## sudo

```
$api->sudo('fred')->create_issue(...);
```

Returns a new instance of [GitLab::API::v4](https://metacpan.org/pod/GitLab%3A%3AAPI%3A%3Av4) with the ["sudo\_user"](#sudo_user) argument
set.

See [https://docs.gitlab.com/ce/api/#sudo](https://docs.gitlab.com/ce/api/#sudo).

# API METHODS

## Award Emoji

See [https://docs.gitlab.com/ce/api/award\_emoji.html](https://docs.gitlab.com/ce/api/award_emoji.html).

- issue\_award\_emojis

    ```perl
    my $award_emojis = $api->issue_award_emojis(
        $project_id,
        $issue_iid,
        \%params,
    );
    ```

    Sends a `GET` request to `projects/:project_id/issues/:issue_iid/award_emoji` and returns the decoded response content.

- merge\_request\_award\_emojis

    ```perl
    my $award_emojis = $api->merge_request_award_emojis(
        $project_id,
        $merge_request_iid,
        \%params,
    );
    ```

    Sends a `GET` request to `projects/:project_id/merge_requests/:merge_request_iid/award_emoji` and returns the decoded response content.

- snippet\_award\_emojis

    ```perl
    my $award_emojis = $api->snippet_award_emojis(
        $project_id,
        $merge_request_id,
        \%params,
    );
    ```

    Sends a `GET` request to `projects/:project_id/merge_requests/:merge_request_id/award_emoji` and returns the decoded response content.

- issue\_award\_emoji

    ```perl
    my $award_emoji = $api->issue_award_emoji(
        $project_id,
        $issue_iid,
        $award_id,
    );
    ```

    Sends a `GET` request to `projects/:project_id/issues/:issue_iid/award_emoji/:award_id` and returns the decoded response content.

- merge\_request\_award\_emoji

    ```perl
    my $award_emoji = $api->merge_request_award_emoji(
        $project_id,
        $merge_request_iid,
        $award_id,
    );
    ```

    Sends a `GET` request to `projects/:project_id/merge_requests/:merge_request_iid/award_emoji/:award_id` and returns the decoded response content.

- snippet\_award\_emoji

    ```perl
    my $award_emoji = $api->snippet_award_emoji(
        $project_id,
        $snippet_id,
        $award_id,
    );
    ```

    Sends a `GET` request to `projects/:project_id/snippets/:snippet_id/award_emoji/:award_id` and returns the decoded response content.

- create\_issue\_award\_emoji

    ```perl
    my $award_emoji = $api->create_issue_award_emoji(
        $project_id,
        $issue_iid,
        \%params,
    );
    ```

    Sends a `POST` request to `projects/:project_id/issues/:issue_iid/award_emoji` and returns the decoded response content.

- create\_merge\_request\_award\_emoji

    ```perl
    my $award_emoji = $api->create_merge_request_award_emoji(
        $project_id,
        $merge_request_iid,
        \%params,
    );
    ```

    Sends a `POST` request to `projects/:project_id/merge_requests/:merge_request_iid/award_emoji` and returns the decoded response content.

- create\_snippet\_award\_emoji

    ```perl
    my $award_emoji = $api->create_snippet_award_emoji(
        $project_id,
        $snippet_id,
    );
    ```

    Sends a `POST` request to `projects/:project_id/snippets/:snippet_id/award_emoji` and returns the decoded response content.

- delete\_issue\_award\_emoji

    ```perl
    my $award_emoji = $api->delete_issue_award_emoji(
        $project_id,
        $issue_id,
        $award_id,
    );
    ```

    Sends a `DELETE` request to `projects/:project_id/issues/:issue_id/award_emoji/:award_id` and returns the decoded response content.

- delete\_merge\_request\_award\_emoji

    ```perl
    my $award_emoji = $api->delete_merge_request_award_emoji(
        $project_id,
        $merge_request_id,
        $award_id,
    );
    ```

    Sends a `DELETE` request to `projects/:project_id/merge_requests/:merge_request_id/award_emoji/:award_id` and returns the decoded response content.

- delete\_snippet\_award\_emoji

    ```perl
    my $award_emoji = $api->delete_snippet_award_emoji(
        $project_id,
        $snippet_id,
        $award_id,
    );
    ```

    Sends a `DELETE` request to `projects/:project_id/snippets/:snippet_id/award_emoji/:award_id` and returns the decoded response content.

- issue\_note\_award\_emojis

    ```perl
    my $award_emojis = $api->issue_note_award_emojis(
        $project_id,
        $issue_iid,
        $note_id,
    );
    ```

    Sends a `GET` request to `projects/:project_id/issues/:issue_iid/notes/:note_id/award_emoji` and returns the decoded response content.

- issue\_note\_award\_emoji

    ```perl
    my $award_emoji = $api->issue_note_award_emoji(
        $project_id,
        $issue_iid,
        $note_id,
        $award_id,
    );
    ```

    Sends a `GET` request to `projects/:project_id/issues/:issue_iid/notes/:note_id/award_emoji/:award_id` and returns the decoded response content.

- create\_issue\_note\_award\_emoji

    ```perl
    my $award_emoji = $api->create_issue_note_award_emoji(
        $project_id,
        $issue_iid,
        $note_id,
        \%params,
    );
    ```

    Sends a `POST` request to `projects/:project_id/issues/:issue_iid/notes/:note_id/award_emoji` and returns the decoded response content.

- delete\_issue\_note\_award\_emoji

    ```perl
    my $award_emoji = $api->delete_issue_note_award_emoji(
        $project_id,
        $issue_iid,
        $note_id,
        $award_id,
    );
    ```

    Sends a `DELETE` request to `projects/:project_id/issues/:issue_iid/notes/:note_id/award_emoji/:award_id` and returns the decoded response content.

- merge\_request\_note\_award\_emojis

    ```perl
    my $award_emojis = $api->merge_request_note_award_emojis(
        $project_id,
        $merge_request_iid,
        $note_id,
    );
    ```

    Sends a `GET` request to `projects/:project_id/merge_requests/:merge_request_iid/notes/:note_id/award_emoji` and returns the decoded response content.

- merge\_request\_note\_award\_emoji

    ```perl
    my $award_emoji = $api->merge_request_note_award_emoji(
        $project_id,
        $merge_request_iid,
        $note_id,
        $award_id,
    );
    ```

    Sends a `GET` request to `projects/:project_id/merge_requests/:merge_request_iid/notes/:note_id/award_emoji/:award_id` and returns the decoded response content.

- create\_merge\_request\_note\_award\_emoji

    ```perl
    my $award_emoji = $api->create_merge_request_note_award_emoji(
        $project_id,
        $merge_request_iid,
        $note_id,
        \%params,
    );
    ```

    Sends a `POST` request to `projects/:project_id/merge_requests/:merge_request_iid/notes/:note_id/award_emoji` and returns the decoded response content.

- delete\_merge\_request\_note\_award\_emoji

    ```perl
    my $award_emoji = $api->delete_merge_request_note_award_emoji(
        $project_id,
        $merge_request_iid,
        $note_id,
        $award_id,
    );
    ```

    Sends a `DELETE` request to `projects/:project_id/merge_requests/:merge_request_iid/notes/:note_id/award_emoji/:award_id` and returns the decoded response content.

## Branches

See [https://docs.gitlab.com/ce/api/branches.html](https://docs.gitlab.com/ce/api/branches.html).

- branches

    ```perl
    my $branches = $api->branches(
        $project_id,
        \%params,
    );
    ```

    Sends a `GET` request to `projects/:project_id/repository/branches` and returns the decoded response content.

- branch

    ```perl
    my $branch = $api->branch(
        $project_id,
        $branch_name,
    );
    ```

    Sends a `GET` request to `projects/:project_id/repository/branches/:branch_name` and returns the decoded response content.

- create\_branch

    ```perl
    my $branch = $api->create_branch(
        $project_id,
        \%params,
    );
    ```

    Sends a `POST` request to `projects/:project_id/repository/branches` and returns the decoded response content.

- delete\_branch

    ```
    $api->delete_branch(
        $project_id,
        $branch_name,
    );
    ```

    Sends a `DELETE` request to `projects/:project_id/repository/branches/:branch_name`.

- delete\_merged\_branches

    ```
    $api->delete_merged_branches(
        $project_id,
    );
    ```

    Sends a `DELETE` request to `projects/:project_id/repository/merged_branches`.

## Broadcast Messages

See [https://docs.gitlab.com/ce/api/broadcast\_messages.html](https://docs.gitlab.com/ce/api/broadcast_messages.html).

- broadcast\_messages

    ```perl
    my $messages = $api->broadcast_messages(
        \%params,
    );
    ```

    Sends a `GET` request to `broadcast_messages` and returns the decoded response content.

- broadcast\_message

    ```perl
    my $message = $api->broadcast_message(
        $message_id,
    );
    ```

    Sends a `GET` request to `broadcast_messages/:message_id` and returns the decoded response content.

- create\_broadcast\_message

    ```perl
    my $message = $api->create_broadcast_message(
        \%params,
    );
    ```

    Sends a `POST` request to `broadcast_messages` and returns the decoded response content.

- edit\_broadcast\_message

    ```perl
    my $message = $api->edit_broadcast_message(
        $message_id,
        \%params,
    );
    ```

    Sends a `PUT` request to `broadcast_messages/:message_id` and returns the decoded response content.

- delete\_broadcast\_message

    ```
    $api->delete_broadcast_message(
        $message_id,
    );
    ```

    Sends a `DELETE` request to `broadcast_messages/:message_id`.

## Project-level Variables

See [https://docs.gitlab.com/ce/api/project\_level\_variables.html](https://docs.gitlab.com/ce/api/project_level_variables.html).

- project\_variables

    ```perl
    my $variables = $api->project_variables(
        $project_id,
        \%params,
    );
    ```

    Sends a `GET` request to `projects/:project_id/variables` and returns the decoded response content.

- project\_variable

    ```perl
    my $variable = $api->project_variable(
        $project_id,
        $variable_key,
    );
    ```

    Sends a `GET` request to `projects/:project_id/variables/:variable_key` and returns the decoded response content.

- create\_project\_variable

    ```perl
    my $variable = $api->create_project_variable(
        $project_id,
        \%params,
    );
    ```

    Sends a `POST` request to `projects/:project_id/variables` and returns the decoded response content.

- edit\_project\_variable

    ```perl
    my $variable = $api->edit_project_variable(
        $project_id,
        $variable_key,
        \%params,
    );
    ```

    Sends a `PUT` request to `projects/:project_id/variables/:variable_key` and returns the decoded response content.

- delete\_project\_variable

    ```
    $api->delete_project_variable(
        $project_id,
        $variable_key,
    );
    ```

    Sends a `DELETE` request to `projects/:project_id/variables/:variable_key`.

## Group-level Variables

See [https://docs.gitlab.com/ce/api/group\_level\_variables.html](https://docs.gitlab.com/ce/api/group_level_variables.html).

- group\_variables

    ```perl
    my $variables = $api->group_variables(
        $group_id,
        \%params,
    );
    ```

    Sends a `GET` request to `groups/:group_id/variables` and returns the decoded response content.

- group\_variable

    ```perl
    my $variable = $api->group_variable(
        $group_id,
        $variable_key,
    );
    ```

    Sends a `GET` request to `groups/:group_id/variables/:variable_key` and returns the decoded response content.

- create\_group\_variable

    ```perl
    my $variable = $api->create_group_variable(
        $group_id,
        \%params,
    );
    ```

    Sends a `POST` request to `groups/:group_id/variables` and returns the decoded response content.

- edit\_group\_variable

    ```perl
    my $variable = $api->edit_group_variable(
        $group_id,
        $variable_key,
        \%params,
    );
    ```

    Sends a `PUT` request to `groups/:group_id/variables/:variable_key` and returns the decoded response content.

- delete\_group\_variable

    ```
    $api->delete_group_variable(
        $group_id,
        $variable_key,
    );
    ```

    Sends a `DELETE` request to `groups/:group_id/variables/:variable_key`.

## Snippets

See [https://docs.gitlab.com/ce/api/snippets.html](https://docs.gitlab.com/ce/api/snippets.html).

- snippets

    ```perl
    my $snippets = $api->snippets();
    ```

    Sends a `GET` request to `snippets` and returns the decoded response content.

- snippet

    ```perl
    my $snippet = $api->snippet(
        $snippet_id,
    );
    ```

    Sends a `GET` request to `snippets/:snippet_id` and returns the decoded response content.

- create\_snippet

    ```perl
    my $snippet = $api->create_snippet(
        \%params,
    );
    ```

    Sends a `POST` request to `snippets` and returns the decoded response content.

- edit\_snippet

    ```perl
    my $snippet = $api->edit_snippet(
        $snippet_id,
        \%params,
    );
    ```

    Sends a `PUT` request to `snippets/:snippet_id` and returns the decoded response content.

- delete\_snippet

    ```
    $api->delete_snippet(
        $snippet_id,
    );
    ```

    Sends a `DELETE` request to `snippets/:snippet_id`.

- public\_snippets

    ```perl
    my $snippets = $api->public_snippets(
        \%params,
    );
    ```

    Sends a `GET` request to `snippets/public` and returns the decoded response content.

- snippet\_user\_agent\_detail

    ```perl
    my $user_agent = $api->snippet_user_agent_detail(
        $snippet_id,
    );
    ```

    Sends a `GET` request to `snippets/:snippet_id/user_agent_detail` and returns the decoded response content.

## Commits

See [https://docs.gitlab.com/ce/api/commits.html](https://docs.gitlab.com/ce/api/commits.html).

- commits

    ```perl
    my $commits = $api->commits(
        $project_id,
        \%params,
    );
    ```

    Sends a `GET` request to `projects/:project_id/repository/commits` and returns the decoded response content.

- create\_commit

    ```perl
    my $commit = $api->create_commit(
        $project_id,
        \%params,
    );
    ```

    Sends a `POST` request to `projects/:project_id/repository/commits` and returns the decoded response content.

- commit

    ```perl
    my $commit = $api->commit(
        $project_id,
        $commit_sha,
    );
    ```

    Sends a `GET` request to `projects/:project_id/repository/commits/:commit_sha` and returns the decoded response content.

- commit\_refs

    ```perl
    my $refs = $api->commit_refs(
        $project_id,
        $commit_sha,
        \%params,
    );
    ```

    Sends a `GET` request to `projects/:project_id/repository/commits/:commit_sha/refs` and returns the decoded response content.

- cherry\_pick\_commit

    ```perl
    my $commit = $api->cherry_pick_commit(
        $project_id,
        $commit_sha,
        \%params,
    );
    ```

    Sends a `POST` request to `projects/:project_id/repository/commits/:commit_sha/cherry_pick` and returns the decoded response content.

- commit\_diff

    ```perl
    my $diff = $api->commit_diff(
        $project_id,
        $commit_sha,
        \%params,
    );
    ```

    Sends a `GET` request to `projects/:project_id/repository/commits/:commit_sha/diff` and returns the decoded response content.

- commit\_comments

    ```perl
    my $comments = $api->commit_comments(
        $project_id,
        $commit_sha,
        \%params,
    );
    ```

    Sends a `GET` request to `projects/:project_id/repository/commits/:commit_sha/comments` and returns the decoded response content.


Changes
Revision history for Perl extension GitLab-API-v4.

{{$NEXT}}

 - Retry on 429 (rate limited) responses.
 - Add a retry_wait with a default of 10 (seconds).

0.27 2023-06-07T20:50:56Z

 - Add run_pipeline_schedule.
 - Add share_group_with_group and unshare_group_with_group.
 - Add approve_merge_request and unapprove_merge_request
 - Add approve_user and reject_user.
 - Add ban_user and unban_user.
 - Add activate_user and deactivate_user.
 - Allow multi-line arguments.
 - Use JSON::MaybeXS instead of JSON.
 - Add verify_SSL=>1 to HTTP::Tiny to verify https server identity.
 - Handle multiple cookies correctly when there is more than one.
 - Live tests now run against GitLab CE 16.0.3 (was 11.10.4).
 - Add the the remote_mirrors, create_remote_mirror, and
   edit_remote_mirror methods.
 - Fixed incorrect license messaging in POD.

0.26 2021-01-30T07:10:57Z

 - Changed licensing terms to be the same as Perl 5.
 - Add the user_memeberships method.
 - Remove file renaming before upload.
 - Provide a .editorconfig instead of .lvimrc.

0.25 2020-02-12T22:20:48Z

 - Add the delete_pipeline method.

0.24 2020-02-12T22:09:39Z

 - Add all the container registry methods.
 - Add http_tiny_request and http_tiny_response to ::RESTClient.  This
   should help people debug issues easier.

0.23 2019-10-18T19:55:06Z

 - Fix delete_runner so that it doesn't try to decode the response.
 - Any 204 response now just blanket returns undef.
 - More contextual JSON decoding error messages.

0.22 2019-09-13T15:02:28Z

 - Add the statistics method.
 - The delete_user method can now take parameters, allowing for use of
   the hard_delete parameter.

0.21 2019-08-24T18:56:13Z

 - The release endpoint returns a 403 on an unknown tag rather than a
   404 like it should.  This condition is now detected and treated
   like a 404.

0.20 2019-07-23T21:42:37Z

 - The preexisting create_release and edit_release methods have been
   renamed to create_tag_release and update_tag_release to make room
   for the new releases API.
 - Added methods for the new (added in GitLab 11.7) releases and
   release links APIs.

0.19 2019-05-17T20:40:22Z

 - Live tests now run against GitLab CE 11.10.4 (was 11.2.3).
 - Add GitLab::API::v4::WWWClient.
 - Add GitLab::API::v4::Mock for writing unit tests.
 - Switch to the GNU General Public License version 3.

0.18 2019-04-01T04:35:02Z

 - Add SUPPORT section to the docs.
 - The archive method no longer tries to decode the response.
 - Add the commit_refs method.
 - Add the trigger_pipeline method.
 - Document some create_pipeline method parameters.

0.17 2019-02-20T22:47:58Z

 - Require perl 5.10.1 as I give up trying to support 5.8 due to other
   modules requiring 5.10 and strictures/multidimensional issues.
 - Simplify VERSION declarations.

0.16 2019-02-18T06:21:15Z

 - Migrate build tooling from Dist::Zilla to Minilla.

0.15 2019-01-09

 - BREAKING CHANGE: gitlab-api-v4 now takes parameters in the form of
   param:value rather than --param=value.  This is to avoid foot-gun
   ambiguities when options have the same names as parameters (such
   as --url).
 - Add --pretty and --canonical options to gitlab-api-v4.

0.14 2018-12-04

 - Add all_project_members and all_group_members methods.
 - gitlab-api-v4 now handles -- like you'd expect your typical linux
   tooling would.

0.13 2018-11-08

 - group() now accepts parameters.
 - Add the transfer_project_to_namespace() method.

0.12 2018-09-11

 - Added API methods for Protected Tags, Search, Project
   Import/Export, Project Badges, Resource Label Events, Discussions,
   Markdown, Group Issue Boards, Group Badges, and Snippets.
 - Reorganized project/group members/access requests sections to
   better match the official API.
 - Renamed the various project snippet methods to have "project" in
   their name.
 - Added constants for the snippet visibility levels.
 - Overhauled the POD heading layout in order to avoid the massive
   TOC.
 - Fix minor typo.

0.11 2018-09-01

 - Fixed the raw_file method to not try to decode the respository file
   contents and instead just return the undecoded response content.
 - Renamed the POD section "FILE METHODS" to
   "REPOSITORY FILE METHODS", to better match the GitLab API docs.
 - Add the hidden NO_ACCESS constant to ::Constants.
 - Live tests now run against GitLab CE 11.2.3 (was 10.2.2).

0.10 2018-06-05

 - Fix project_hook() so it uses the correct path.
 - create_project_hook() and edit_project_hook() now return their
   decoded response (the hook).
 - delete_project_hook() no longer tries to return anything as a
   successful delete is a 204 (no content) and a failed one is a 404.
 - Tests for all of the above.
 - Fix typo.

0.09 2018-05-28

 - Ensure chmod 600 of ~/.gitlab-api-v4-config is maintained.
 - Enable and fix upload_file_to_project by using
   HTTP::Tiny::Multipart.
 - Individual methods may now have a custom note added to their POD.
 - Added a self-contained unit test, unit.t, which can be expanded on
   in the future.  This is unlike regression.t which requires that a
   GitLab installation be available.
 - Added the rest_client_class argument.

0.08 2018-05-14

 - Many API endpoints which return lists and did not support
   parameters now do in order for pagination arguments, and the
   paginator, to be used with them.
 - Add project_languages method.
 - Remove upload_file_to_project since it doesn't work.  Will re-add
   when it is made to work later.

0.07 2018-05-10

 - Completely overhaul ::RestClient to use HTTP::Tiny instead of
   Role::REST::Client.  This change will make it much easier to alter
   the behavior of the HTTP communication.
 - ::RestClient now supports the ability to decode non-ref JSON, such
   as unblock_user which returns just a JSON boolean.
 - block_user and unblock_user now return their decoded response.
 - delete_user is no longer expected to return anything.
 - Response decoding will now only be attempted for methods which are

author/header.pm
package GitLab::API::v4;
our $VERSION = '0.26';

=encoding utf8

=head1 NAME

GitLab::API::v4 - A complete GitLab API v4 client.

=head1 SYNOPSIS

    use GitLab::API::v4;
    
    my $api = GitLab::API::v4->new(
        url           => $v4_api_url,
        private_token => $token,
    );
    
    my $branches = $api->branches( $project_id );

=head1 DESCRIPTION

This module provides a one-to-one interface with the GitLab
API v4.  Much is not documented here as it would just be duplicating
GitLab's own L<API Documentation|http://doc.gitlab.com/ce/api/README.html>.

Note that this distribution also includes the L<gitlab-api-v4> command-line
interface (CLI).

=head2 Upgrading

If you are upgrading from L<GitLab::API::v3> make sure you read:

L<https://docs.gitlab.com/ce/api/v3_to_v4.html>

Also, review the C<Changes> file included in the distribution as it outlines
the changes made to convert the v3 module to v4:

L<https://github.com/bluefeet/GitLab-API-v4/blob/master/Changes>

Finally, be aware that many methods were added, removed, renamed, and/or altered.
If you want to review exactly what was changed you can use GitHub's compare tool:

L<https://github.com/bluefeet/GitLab-API-v4/compare/72e384775c9570f60f8ef68dee3a1eecd347fb69...master>

Or clone the repo and run this command:

C<git diff 72e384775c9570f60f8ef68dee3a1eecd347fb69..HEAD -- author/sections/>

=head2 Credentials

Authentication credentials may be defined by setting either the L</access_token>
or L</private_token> arguments.

If no credentials are supplied then the client will be anonymous and greatly
limited in what it can do with the API.

Extra care has been taken to hide the token arguments behind closures.  This way,
if you dump your api object, your tokens won't accidentally leak into places you
don't want them to.

=head2 Constants

The GitLab API, in rare cases, uses a hard-coded value to represent a state.
To make life easier the L<GitLab::API::v4::Constants> module exposes
these states as named variables.

=head2 Exceptions

The API methods will all throw a useful exception if
an unsuccessful response is received from the API.  That is except for
C<GET> requests that return a C<404> response - these will return C<undef>
for methods that return a value.

If you'd like to catch and handle these exceptions consider using
L<Try::Tiny>.

=head2 Logging

This module uses L<Log::Any> and produces some debug messages here
and there, but the most useful bits are the info messages produced
just before each API call.

=head2 Project ID

Note that many API calls require a C<$project_id>.  This can be
specified as a numeric project C<ID> or, in many cases, maybe all cases,
as a C<NAMESPACE_PATH/PROJECT_PATH> string.  The GitLab documentation on
this point is vague.

=cut

use Carp qw( croak );
use GitLab::API::v4::Paginator;
use GitLab::API::v4::RESTClient;
use Log::Any qw( $log );
use Types::Common::Numeric -types;
use Types::Common::String -types;
use Types::Standard -types;

use Moo;
use strictures 2;
use namespace::clean;

sub BUILD {
    my ($self) = @_;

    # Ensure any token arguments get moved into their closure before we return
    # the built object.
    $self->access_token();
    $self->private_token();

    $log->debugf( "An instance of %s has been created.", ref($self) );

    return;
}

sub _call_rest_client {
    my ($self, $verb, $path, $path_vars, $options) = @_;

    $options->{headers} = $self->_auth_headers();

    return $self->rest_client->request(
        $verb, $path, $path_vars, $options,
    );
}

sub _auth_headers {
    my ($self) = @_;
    my $headers = {};

    $headers->{'authorization'} = 'Bearer ' . $self->access_token()
        if defined $self->access_token();
    $headers->{'private-token'} = $self->private_token()
        if defined $self->private_token();
    $headers->{'sudo'} = $self->sudo_user()
        if defined $self->sudo_user();

    return $headers;
}

sub _clone_args {
    my ($self) = @_;

    return {
        url         => $self->url(),
        retries     => $self->retries(),
        rest_client => $self->rest_client(),
        (defined $self->access_token()) ? (access_token=>$self->access_token()) : (),
        (defined $self->private_token()) ? (private_token=>$self->private_token()) : (),
    };
}

sub _clone {
    my $self = shift;

    my $class = ref $self;
    my $args = {
        %{ $self->_clone_args() },
        %{ $class->BUILDARGS( @_ ) },
    };

    return $class->new( $args );
}

# Little utility method that avoids any ambiguity in whether a closer is
# causing circular references.  Don't ever pass it a ref.
sub _make_safe_closure {
    my ($ret) = @_;
    return sub{ $ret };
}

=head1 REQUIRED ARGUMENTS

=head2 url

The URL to your v4 API endpoint.  Typically this will be something
like C<https://git.example.com/api/v4>.

=cut

has url => (
    is       => 'ro',
    isa      => NonEmptySimpleStr,
    required => 1,
);

=head1 OPTIONAL ARGUMENTS

=head2 access_token

A GitLab API OAuth2 token.  If set then L</private_token> may not be set.

See L<https://docs.gitlab.com/ce/api/#oauth2-tokens>.

=cut

has _access_token_arg => (
    is        => 'ro',
    isa       => NonEmptySimpleStr,
    init_arg  => 'access_token',
    clearer   => '_clear_access_token_arg',
);

has _access_token_closure => (
    is       => 'lazy',
    isa      => CodeRef,
    init_arg => undef,
    builder  => '_build_access_token_closure',
);
sub _build_access_token_closure {
    my ($self) = @_;
    my $token = $self->_access_token_arg();
    $self->_clear_access_token_arg();
    return _make_safe_closure( $token );
}

sub access_token {
    my ($self) = @_;
    return $self->_access_token_closure->();
}

=head2 private_token

A GitLab API personal token.  If set then L</access_token> may not be set.

See L<https://docs.gitlab.com/ce/api/#personal-access-tokens>.

=cut

has _private_token_arg => (
    is        => 'ro',
    isa       => NonEmptySimpleStr,
    init_arg  => 'private_token',
    clearer   => '_clear_private_token_arg',
);

has _private_token_closure => (
    is       => 'lazy',
    isa      => CodeRef,
    init_arg => undef,
    builder  => '_build_private_token_closure',
);
sub _build_private_token_closure {
    my ($self) = @_;
    my $token = $self->_private_token_arg();
    $self->_clear_private_token_arg();
    return _make_safe_closure( $token );
}

sub private_token {
    my ($self) = @_;
    return $self->_private_token_closure->();
}

=head2 retries

The number of times the request should be retried in case it fails (5XX HTTP
response code).  Defaults to C<0> (false), meaning that a failed request will
not be retried.

=cut

has retries => (
    is      => 'ro',
    isa     => PositiveOrZeroInt,
    default => 0,
);

=head2 sudo_user

The user to execute API calls as.  You may find it more useful to use the
L</sudo> method instead.

See L<https://docs.gitlab.com/ce/api/#sudo>.

=cut

has sudo_user => (
    is  => 'ro',
    isa => NonEmptySimpleStr,
);

=head2 rest_client

An instance of L<GitLab::API::v4::RESTClient> (or whatever L</rest_client_class>
is set to).  Typically you will not be setting this as it defaults to a new
instance and customization should not be necessary.

=cut

has rest_client => (
    is  => 'lazy',
    isa => InstanceOf[ 'GitLab::API::v4::RESTClient' ],
);
sub _build_rest_client {
    my ($self) = @_;

    return $self->rest_client_class->new(
        base_url => $self->url(),
        retries  => $self->retries(),
    );
}

=head2 rest_client_class

The class to use when constructing the L</rest_client>.
Defaults to L<GitLab::API::v4::RESTClient>.

=cut

has rest_client_class => (
    is  => 'lazy',
    isa => NonEmptySimpleStr,
);
sub _build_rest_client_class {
    return 'GitLab::API::v4::RESTClient';
}

=head1 UTILITY METHODS

=head2 paginator

    my $paginator = $api->paginator( $method, @method_args );
    
    my $members = $api->paginator('group_members', $group_id);
    while (my $member = $members->next()) {
        ...
    }
    
    my $users_pager = $api->paginator('users');
    while (my $users = $users_pager->next_page()) {
        ...
    }
    
    my $all_open_issues = $api->paginator(
        'issues',
        $project_id,
        { state=>'opened' },
    )->all();

Given a method who supports the C<page> and C<per_page> parameters,
and returns an array ref, this will return a L<GitLab::API::v4::Paginator>
object that will allow you to walk the records one page or one record
at a time.

=cut

sub paginator {
    my ($self, $method, @args) = @_;

    my $params = (ref($args[-1]) eq 'HASH') ? pop(@args) : {};

    return GitLab::API::v4::Paginator->new(
        api    => $self,
        method => $method,
        args   => \@args,
        params => $params,
    );
}

=head2 sudo

    $api->sudo('fred')->create_issue(...);

Returns a new instance of L<GitLab::API::v4> with the L</sudo_user> argument
set.

See L<https://docs.gitlab.com/ce/api/#sudo>.

=cut

sub sudo {
    my ($self, $user) = @_;

    return $self->_clone(
        sudo_user => $user,
    );
}

